File: go/internal/e2e/hooks_extended_e2e_test.go
package e2e

import (
	"strings"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func assertUnsupportedExtendedHooks(t *testing.T, client *copilot.Client, hooks *copilot.SessionHooks) {
	t.Helper()

	session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		Hooks:               hooks,
	})
	if err == nil {
		if session != nil {
			_ = session.Disconnect()
		}
		t.Fatal("expected SDK callback hooks to be rejected")
	}
	if !strings.Contains(err.Error(), unsupportedSDKHooksMessage) {
		t.Fatalf("expected unsupported hooks error, got %v", err)
	}
}

func TestHooksExtendedE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	cases := map[string]*copilot.SessionHooks{
		"userPromptSubmitted": {
			OnUserPromptSubmitted: func(copilot.UserPromptSubmittedHookInput, copilot.HookInvocation) (*copilot.UserPromptSubmittedHookOutput, error) {
				return &copilot.UserPromptSubmittedHookOutput{ModifiedPrompt: "not used"}, nil
			},
		},
		"sessionStart": {
			OnSessionStart: func(copilot.SessionStartHookInput, copilot.HookInvocation) (*copilot.SessionStartHookOutput, error) {
				return &copilot.SessionStartHookOutput{AdditionalContext: "not used"}, nil
			},
		},
		"sessionEnd": {
			OnSessionEnd: func(copilot.SessionEndHookInput, copilot.HookInvocation) (*copilot.SessionEndHookOutput, error) {
				return &copilot.SessionEndHookOutput{SessionSummary: "not used"}, nil
			},
		},
		"errorOccurred": {
			OnErrorOccurred: func(copilot.ErrorOccurredHookInput, copilot.HookInvocation) (*copilot.ErrorOccurredHookOutput, error) {
				return &copilot.ErrorOccurredHookOutput{ErrorHandling: "skip"}, nil
			},
		},
		"preToolUse output": {
			OnPreToolUse: func(copilot.PreToolUseHookInput, copilot.HookInvocation) (*copilot.PreToolUseHookOutput, error) {
				return &copilot.PreToolUseHookOutput{PermissionDecision: "allow"}, nil
			},
		},
		"postToolUse output": {
			OnPostToolUse: func(copilot.PostToolUseHookInput, copilot.HookInvocation) (*copilot.PostToolUseHookOutput, error) {
				return &copilot.PostToolUseHookOutput{SuppressOutput: false}, nil
			},
		},
		"postToolUseFailure output": {
			OnPostToolUse: func(copilot.PostToolUseHookInput, copilot.HookInvocation) (*copilot.PostToolUseHookOutput, error) {
				return nil, nil
			},
			OnPostToolUseFailure: func(copilot.PostToolUseFailureHookInput, copilot.HookInvocation) (*copilot.PostToolUseFailureHookOutput, error) {
				return &copilot.PostToolUseFailureHookOutput{AdditionalContext: "not used"}, nil
			},
		},
	}

	for name, hooks := range cases {
		t.Run("rejects SDK callback hook "+name, func(t *testing.T) {
			ctx.ConfigureForTest(t)
			assertUnsupportedExtendedHooks(t, client, hooks)
		})
	}
}

File: nodejs/test/e2e/hooks_extended.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import type { SessionHooks } from "../../src/types.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

const UNSUPPORTED_SDK_HOOKS_MESSAGE = "SDK hook callbacks are no longer supported";

describe("Extended session hooks", async () => {
    const { copilotClient: client } = await createSdkTestContext();

    async function expectUnsupportedHooks(hooks: SessionHooks) {
        await expect(
            client.createSession({
                onPermissionRequest: approveAll,
                hooks,
            })
        ).rejects.toThrow(UNSUPPORTED_SDK_HOOKS_MESSAGE);
    }

    const hookCases: Array<[string, SessionHooks]> = [
        ["userPromptSubmitted", { onUserPromptSubmitted: () => undefined }],
        ["sessionStart", { onSessionStart: () => undefined }],
        ["sessionEnd", { onSessionEnd: () => undefined }],
        ["errorOccurred", { onErrorOccurred: () => undefined }],
        ["preToolUse output", { onPreToolUse: () => ({ permissionDecision: "allow" }) }],
        ["postToolUse output", { onPostToolUse: () => ({ suppressOutput: false }) }],
        [
            "postToolUseFailure output",
            {
                onPostToolUse: () => undefined,
                onPostToolUseFailure: () => ({ additionalContext: "not used" }),
            },
        ],
    ];

    it.each(hookCases)("rejects SDK callback hook %s", async (_name, hooks) => {
        await expectUnsupportedHooks(hooks);
    });
});
